### PR TITLE
Add file based locking to TBot

### LINKED ISSUE #17794: `tbot` should lock its data directory to prevent multiple bots from starting concurrently
What would you like Teleport to do?

`tbot` should lock its internal data directory (and possibly its destination directories as well, though this one is less vital)

What problem does this solve?

`tbot` processes are very stateful and can trigger various unpleasant conditions if multiple processes are competing for the same internal data directory:
 * one bot may consume join tokens and cause subsequent runs to fail
 * a running bot can invalidate the certificates of another bot, causing a generation counter lockout

It's not very difficult to accidentally start multiple bots with the same config, especially when run in the background (e.g. systemd unit).

If a workaround exists, please include it.

Don't start multiple `tbot` instances at the same time with the same data dir

### PR DESCRIPTION
**Pull request by strideynet**
Leverages the existing `flock` package used elsewhere in Teleport to ensure only one `tbot` instance can be running against a data directory at one time.

Closes https://github.com/gravitational/teleport/issues/17794

image

### CODE REVIEW

**Review by timothyb89 — APPROVED** [review]
This works and fixes the issue! My only concern (nitpick?) is that while the API provides `Lock()` for destination dirs (as well as the storage dir), locking could fail on destination dirs that are using ACLs with `tbot init` since there won't be any permissions to access the lockfile.

I'm not sure if ACLs and write locks play nice together, but since we don't ever attempt to lock destination dirs, and `tbot init` doesn't support storage dirs anyway, maybe just a TODO/warning/etc is good enough for now?

File: lib/tbot/tbot.go [hunk 1]
```diff
@@ -224,7 +225,15 @@ func (b *Bot) Run(ctx context.Context) error {
 		return trace.Wrap(err)
 	}
 
-	if err := b.initialize(ctx); err != nil {
+	unlock, err := b.initialize(ctx)
+	defer func() {
+		if unlock != nil {
```
Comment (camscale) [inline lib/tbot/tbot.go:230]: if you moved the defer after the error check, you should not need this nil check

File: lib/tbot/tbot.go [hunk 2]
```diff
@@ -224,7 +225,15 @@ func (b *Bot) Run(ctx context.Context) error {
...
+			if err := unlock(); err != nil {
+				b.log.WithError(err).Warn("Failed to release lock. Future starts of tbot may fail.")
```
Comment (camscale) [inline lib/tbot/tbot.go:232]: Is this true? When this tbot process exits, the lock should be released by the kernel (according to flock(2)). I wonder if this log message may instil unnecessary worry in users.

**Review by camscale — APPROVED** [review]
LGTM. Just some opinion-based comments. feel free to ignore.

File: lib/tbot/tbot.go (hunk 1 above)
Comment (strideynet) [reply lib/tbot/tbot.go:230]: If I move the defer after the error check, there's a possibility a lock will not be released if another error occurs inside `initialize`.

File: lib/tbot/tbot.go (hunk 2 above)
Comment (strideynet) [reply lib/tbot/tbot.go:232]: This is certainly true for `flock` - however, we are looking into supporting other destinations such as Kubernetes secrets etc, and for these, the automagic behaviour of `flock` will probably not necessarily apply.

File: lib/tbot/bot/destination.go [hunk 3]
```diff
@@ -33,4 +33,8 @@ type Destination interface {
 
 	// Read fetches data from the destination with a given name.
 	Read(name string) ([]byte, error)
+
+	// Lock attempts to lock a destination. This is used so we can prevent
+	// two bots operating against the same directory or datastore concurrently.
+	Lock() (func() error, error)
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/tbot/bot/destination.go]: 6538944a0c

**Comment by strideynet** [comment]
> This works and fixes the issue! My only concern (nitpick?) is that while the API provides `Lock()` for destination dirs (as well as the storage dir), locking could fail on destination dirs that are using ACLs with `tbot init` since there won't be any permissions to access the lockfile.
> 
> I'm not sure if ACLs and write locks play nice together, but since we don't ever attempt to lock destination dirs, and `tbot init` doesn't support storage dirs anyway, maybe just a TODO/warning/etc is good enough for now?

Yeah, this is a good point, I'll add an appropriate warning for now and we can revisit this at a later date if we decide to lock destination dirs.

**Pull request merged**